llvm/lib/SYCL/KernelProperties.cpp

```diff
@@ -10,72 +10,125 @@
+               ->getInitializer());
+      unsigned Bank;
+      if (auto *ZeroData = dyn_cast<ConstantAggregateZero>(Args))
+        Bank = 0;
+      else
+        Bank = cast<ConstantInt>(Args->getOperand(0))->getZExtValue();
 
-    userSpecifiedDDRBanks[Alloca] = Bank;
-    if (maxiBundles.find(Bank) == maxiBundles.end()) {
-      maxiBundles[Bank] = {formatv("ddrmem{0}", Bank)};
+      UserSpecifiedDDRBanks[Alloca] = Bank;
     }
   }
-}
 
-Optional<unsigned> KernelProperties::getUserSpecifiedDDRBank(Argument *Arg) {
+  Optional<unsigned> getUserSpecifiedDDRBank(Argument *Arg, SmallDenseMap<llvm::AllocaInst *, unsigned, 16> &UserSpecifiedDDRBanks) {
   for (User *U : Arg->users()) {
     if (auto *Store = dyn_cast<StoreInst>(U))
       if (Store->getValueOperand() == Arg) {
-        auto Lookup = userSpecifiedDDRBanks.find(dyn_cast_or_null<AllocaInst>(
+        auto Lookup = UserSpecifiedDDRBanks.find(dyn_cast_or_null<AllocaInst>(
             getUnderlyingObject(Store->getPointerOperand())));
-        if (Lookup == userSpecifiedDDRBanks.end())
+        if (Lookup == UserSpecifiedDDRBanks.end())
           continue;
         return {Lookup->second};
       }
   }
   return {};
 }
+} // namespace
 
-Optional<StringRef> KernelProperties::getArgumentMAXIBundle(Argument *Arg) {
-  auto DdrId = getUserSpecifiedDDRBank(Arg);
-  if (DdrId) {
-    return {maxiBundles[DdrId.getValue()].bundleName};
+namespace llvm {
+bool KernelProperties::isArgBuffer(Argument* Arg, bool SyclHLSFlow) {
+  if (Arg->getType()->isPointerTy() &&
+      (SyclHLSFlow ||
+       Arg->getType()->getPointerAddressSpace() == SPIRAS_Global ||
+       Arg->getType()->getPointerAddressSpace() == SPIRAS_Constant)) {
+    return !Arg->hasByValAttr();
   }
-  return {};
+  return false;
+}
+
+KernelProperties::KernelProperties(Function &F, bool SyclHlsFlow) {
+
+  SmallDenseMap<llvm::AllocaInst*, unsigned, 16> UserSpecifiedDDRBanks{};
+  collectUserSpecifiedDDRBanks(F, UserSpecifiedDDRBanks);
+
+  SmallDenseMap<Argument *, unsigned, 16> DDRAssignment{};
+  for (auto &Arg : F.args()) {
+    if (isArgBuffer(&Arg, SyclHlsFlow)) {
+           auto Assignment = getUserSpecifiedDDRBank(&Arg, UserSpecifiedDDRBanks);
+           DDRAssignment[&Arg] = (Assignment) ? Assignment.getValue() : 0;
+         }
+  }
+
+  for (auto &ArgBank : DDRAssignment) {
+    auto LookUp = BundlesByIDName.find(ArgBank.second);
+    if (LookUp == BundlesByIDName.end()) {
+      // We need to create a bundle for this bank
+      auto BundleName = formatv("ddrmem{0}", ArgBank.second);
+      Bundles.push_back({ BundleName, ArgBank.second});
+      unsigned BundleIdx = Bundles.size() - 1;
+      BundlesByName[ {BundleName} ] = BundleIdx;
+      BundleForArgument[ArgBank.first] = BundleIdx;
+      BundlesByIDName[ArgBank.second][ {BundleName} ] = BundleIdx;
+    } else {
+      BundleForArgument[ArgBank.first] = LookUp->getSecond().begin()->second;
```

Comment:
There is no C++20 modern equivalent syntax for the LLVM maps for the code above?

Reply:
What would you suggest ?